### PR TITLE
Detection of bundled CMake in Visual Studio

### PR DESCRIPTION
**Pull request by Zingam**
## This change addresses item #610

The following changes are proposed:

- Moved some *common* functions into a new package 'installs/visual-studio'.

## The purpose of this change

Enables the detection of bundled CMake packages in Visual Studio.

## Other Notes/Information

A future improvement could be to scan the system once or on demand for all CMake versions and let the user pick one via the UI. When 'cmake.cmakePath' is not set  the extension appears to query the system for CMake executables continuously.

### CODE REVIEW

File: src/paths.ts [hunk 1]
```diff
@@ -128,19 +130,51 @@ class Paths {
...
+            for (const tool_path of bundled_tool_paths) {
+              if (await fs.exists(tool_path.cmake)) {
+                // Append the bundled Ninja build system path to the VSCode's process' environment variable 'PATH'
+                if (await fs.exists(tool_path.ninja)) {
+                  if (process.env.hasOwnProperty('PATH')) {
+                    const env_paths = (process.env.PATH as string).split(';');
+                    const ninja_path = path.dirname(tool_path.ninja);
+                    const ninja_base_path = env_paths.find(path_el => path_el === ninja_path);
+                    if (ninja_base_path === undefined) {
+                      (process.env.PATH as string) = (process.env.PATH as string).concat(';' + ninja_path);
```
Comment (bobbrow) [inline src/paths.ts]: Changing the process.env only works inside the node process and I wouldn't expect it to work automatically for spawned processes like CMake.  Can you split this out into a new function to get the Ninja path that we can call from `getVSKitEnvironment` instead?

**Comment by Zingam** [comment]
@bobbrow I'll wait for Multi-Root, File API, etc to be merged before I go over my current PRs again.

**Comment by bobbrow** [comment]
Multi-root is in.  I don't think this will conflict with file-api.

File: src/installs/visual-studio.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
+/**
+ * Module for querying MS Visual Studio
```
[the remark this answers is not in the document]
Comment (Zingam) [reply src/installs/visual-studio.ts]: Are these concerns still valid?

[the remark this answers is not in the document]
Comment (bobbrow) [reply src/installs/visual-studio.ts]: It looks like it merged cleanly, but I don't think you should bring as much stuff in here.  Really, this file would just be about querying the installation information.  Some of the other stuff in here is more related to Kits.  I'll mark the functions that I think should move back to `kit.ts`.

File: src/installs/visual-studio.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import * as logging from '../logging';
+import * as proc from '../proc';
+import {thisExtensionPath} from '../util';
+import * as nls from 'vscode-nls';
+
+
+nls.config({ messageFormat: nls.MessageFormat.bundle, bundleFormat: nls.BundleFormat.standalone })();
+const localize: nls.LocalizeFunc = nls.loadMessageBundle();
+
+const log = logging.createLogger('kit');
```
Comment (bobbrow) [inline src/installs/visual-studio.ts]: I think this logger should have a different name.

File: src/installs/visual-studio.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+let cachedVSInstallations: VSInstallationCache|null = null;
+
+
+/**
+ * Construct the display name (this will be paired with an
+ * arch later to construct the Kit.name property).
+ *
+ * @param inst The VSInstallation to use
+ */
+export function vsDisplayName(inst: VSInstallation): string {
```
Comment (bobbrow) [inline src/installs/visual-studio.ts]: This should go back to `kit.ts`

File: src/installs/visual-studio.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    }
+  }
+  cachedVSInstallations = {
+    installations: installs,
+    queryTime: now
+  };
+  return installs;
+}
+
+export function vsVersionName(inst: VSInstallation): string {
```
Comment (bobbrow) [inline src/installs/visual-studio.ts]: This should go back to `kit.ts`

**Comment by Zingam** [comment]
I hope this initial version is OK for now. I plan to implement a more sophisticated CMake detection later -> https://github.com/microsoft/vscode-cmake-tools/issues/943#issuecomment-564397901

File: src/paths.ts [hunk 6]
```diff
@@ -128,28 +142,70 @@ class Paths {
...
+      }
+
+      for (const tool_path of bundled_tool_paths) {
+        if (await fs.exists(tool_path.cmake)) {
+          // Append the bundled Ninja build system path to the VSCode's process' environment variable 'PATH'
+          if (await fs.exists(tool_path.ninja)) {
+            vsCMakePaths.ninja = tool_path.ninja;
+          }
+          // CMake can be still used without Ninja
+          vsCMakePaths.cmake = tool_path.cmake;
```
Comment (bobbrow) [inline src/paths.ts]: Should this break out of the loop once you find one that works?

Comment (Zingam) [reply src/paths.ts]: Yes, that was the intention. Thank you.

**Pull request merged**